[colmap] Add port for COLMAP 3.6

pablospe opened the pull request:
Add port for COLMAP 3.6-dev.3. Related to issue #8820.

pablospe committed:
[colmap] Add port for COLMAP 3.6-dev.3

pablospe committed:
Fixed some errors in portfile

pablospe committed:
Fixing some errors in portfile. These changes also move the binary (and *.bat) to `tools/` otherwise I get `POST_BUILD_CHECKS_FAILED`. I followed this recommendation:
https://github.com/Microsoft/vcpkg/issues/834#issuecomment-288529839

Now the *.bat files need to be fixed with the correct path to `tools/`

pablospe commented:
Last commit fixed some errors in portfile. These changes also move the binary (and *.bat) to `tools/` folder, otherwise I get `POST_BUILD_CHECKS_FAILED` with these errors:

```
The following EXEs were found in /bin or /debug/bin. EXEs are not valid distribution targets.
    C:/build/vcpkg/packages/colmap_x64-windows/bin/colmap.exe
The following EXEs were found in /bin or /debug/bin. EXEs are not valid distribution targets.
    C:/build/vcpkg/packages/colmap_x64-windows/debug/bin/colmap.exe
Found 2 error(s). Please correct the portfile
```

I followed this recommendation and moved it to `tools` folder: https://github.com/Microsoft/vcpkg/issues/834#issuecomment-288529839
But now the `*.bat` files need to be fixed with the correct path to `tools/`

A second problem is that when I run the binary I get: _**The code execution cannot proceed because `boost_filesystem-vc142-mt-x64-1_73.dll` was not found**_. This doesn't happen to me if I compile colmap with vcpkg following [these instructions](https://github.com/microsoft/vcpkg/issues/8820#issuecomment-657275423), perhaps the `install` target is the difference. I was playing with the RPATH (see [this link](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling#always-full-rpath)) to see if I could fixed but it didn't make any difference.

ahojnnes commented on ports/colmap/portfile.cmake:
```diff
@@ -0,0 +1,64 @@
+
+vcpkg_fixup_cmake_targets()
+
+file(REMOVE_RECURSE
```
Why do we remove the headers?

pablospe replied:
Because I got the following errors, and without removing these post-build validation
I doesn't pass (POST_BUILD_CHECKS_FAILED)

```
-- Performing post-build validation
Include files should not be duplicated into the /debug/include directory. If this cannot be disabled in the project cmake, use
    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
/debug/share should not exist. Please reorganize any important files, then use
    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
```

I based on the libogg example ([doc ](https://github.com/microsoft/vcpkg/blob/master/docs/examples/packaging-github-repos.md) and [portfile](https://github.com/microsoft/vcpkg/blob/master/ports/libogg/portfile.cmake))

ras0219-msft commented on ports/colmap/portfile.cmake:
```diff
@@ -0,0 +1,64 @@
+    "${CURRENT_PACKAGES_DIR}/include/colmap/ui/shaders"
+    "${CURRENT_PACKAGES_DIR}/debug/COLMAP.bat"
+    "${CURRENT_PACKAGES_DIR}/debug/RUN_TESTS.bat"
+    "${CURRENT_PACKAGES_DIR}/debug/bin"
```
This line should no longer be needed with `vcpkg_copy_tools(TOOL_NAMES colmap AUTO_CLEAN)`.

pablospe replied:
I still need to remove the empty folders

ras0219-msft commented on ports/colmap/portfile.cmake:
```diff
@@ -0,0 +1,64 @@
+vcpkg_from_github(
```
It looks like this library may only support static libs -- in this case, we need to add
```
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```
to the top of the portfile.

pablospe replied:
Shouldn't this be controlled by the `--tritple` option? `--triplet x64-windows` vs `--triplet x64-windows-static`

pablospe replied:
With the latest commints `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)` is not needed.

ras0219-msft reviewed (changes requested):
Thanks for the PR!

pablospe committed:
Update ports/colmap/portfile.cmake

Co-authored-by: Robert Schumacher <roschuma@microsoft.com>

pablospe committed:
Update ports/colmap/portfile.cmake

Co-authored-by: Robert Schumacher <roschuma@microsoft.com>

pablospe committed:
Update ports/colmap/usage

Co-authored-by: Robert Schumacher <roschuma@microsoft.com>

pablospe commented:
@ras0219-msft, here you were suggesting to use `vcpkg_copy_tools()`:
https://github.com/microsoft/vcpkg/pull/12410#discussion_r455387065

but the `sqlite` example doesn't use `vcpkg_copy_tools()`. Are these the lines you refer to?
```
vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
    INVERTED_FEATURES
    tool SQLITE3_SKIP_TOOLS
)
```

I think this will require some changes in the COLMAP code, to avoid compiling the binaries tools using a variable that can be disable here. Anyways I would assume that the regular user of COLMAP would expect the binaries to be compiled by default.

pablospe committed:
First working version, the line `vcpkg_copy_tools(TOOL_NAMES colmap AUTO_CLEAN)` helped.

Now it does work running:

    > <vcpkg-root>\vcpkg\packages\colmap_x64-windows\tools\colmap\colmap.exe
    > <vcpkg-root>\vcpkg\packages\colmap_x64-windows\tools\colmap\colmap.exe gui

ToDo: use `vcpkg.json`.

pablospe committed:
Using `vcpkg.json` instead of CONTROL file

pablospe commented:
Please try:

    <vcpkg-root>\vcpkg.exe install colmap[cuda] --triplet x64-windows

Now it should work:

    <vcpkg-root>\vcpkg\packages\colmap_x64-windows\tools\colmap\colmap.exe
    <vcpkg-root>\vcpkg\packages\colmap_x64-windows\tools\colmap\colmap.exe gui

pablospe committed:
Formatting json file:
./vcpkg.exe x-format-manifest --all

pablospe reviewed (commented):
Most of the changes were incorporated, except that I still need to remove some empty folders

pablospe commented:
qt5-imageformats:x86-windows fails to build. Any idea what to do in these cases? For the rest architectures seems to work (x64_windows, x64_linux, etc.).

[qt5-imageformats:x86-windows: BUILD_FAILED](https://dev.azure.com/vcpkg/public/_build/results?buildId=40421&view=logs&j=5dffab7c-3299-5a76-7b02-e037cd868fad&t=70564965-7ae4-5af3-8764-a93ab8ddf245&l=5630)

ahojnnes commented:
@pablospe I ran into a few issues compiling this on my machine.

I made a few commits yesterday on the COLMAP side to fix them.

In addition, I suggest to incorporate the following improvements:

```
if("cuda-redist" IN_LIST FEATURES)
    set(CUDA_ENABLED ON)
    set(CUDA_ARCHS "Common")
endif()

vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DCUDA_ENABLED=${CUDA_ENABLED}
        -DCUDA_ARCHS=${CUDA_ARCHS}
)
```

```
  "features": [
    {
      "name": "cuda",
      "dependencies": [
        "cuda"
      ],
      "description": "CUDA support for current compute architecture of this machine"
    },
    {
      "name": "cuda-redist",
      "dependencies": [
        "cuda"
      ],
      "description": "Redistributable CUDA support for common supported compute architectures"
    }
  ]
```

ahojnnes commented:
The cuda-redist feature allows one to compile a redistributable COLMAP version with CUDA compiled to different GPU architectures.

ahojnnes commented:
In addition, do you know, if there is a way to tell vcpkg to compile the latest commit from a specific branch from Github instead of checking out the predefined commit here?

ahojnnes commented:
Another comment: COLMAP automatically extracts version information from the git log at compile time using `cmake/GenerateVersionDefinitions.cmake`. It's not super important, but at the moment, this will produce a "Unknown" version number.

ahojnnes commented:
> In addition, do you know, if there is a way to tell vcpkg to compile the latest commit from a specific branch from Github instead of checking out the predefined commit here?

Okay, answered this myself. Can be done with `install --head`.

pablospe committed:
Changes suggested by @ahojnnes (colmap author). Waiting for the 3.7 release  (probably today) to update the `REF` and `SHA512`

pablospe commented:
> Another comment: COLMAP automatically extracts version information from the git log at compile time using `cmake/GenerateVersionDefinitions.cmake`. It's not super important, but at the moment, this will produce a "Unknown" version number.

To know the `GIT_COMMIT_ID ` in [colmap/cmake/GenerateVersionDefinitions.cmake](https://github.com/colmap/colmap/blob/dev/cmake/GenerateVersionDefinitions.cmake) it is possible to use "-D..." option with `$VCPKG_HEAD_VERSION` variable. But I don't know if there is a variable for the `GIT_COMMIT_DATE`. Any idea? @ras0219-msft, @JackBoosY.

pablospe commented:
I was trying to compile with `x64-windows-static` triplet and I got these kind of errors:

```
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:136 (message):
    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v
    Working Directory: C:/build/vcpkg/buildtrees/colmap/x64-windows-static-dbg
    See logs for more information:
      C:\build\vcpkg\buildtrees\colmap\install-x64-windows-static-dbg-out.log
```

And in the file `install-x64-windows-static-dbg-out.log`:

```
FreeImage.lib(BitmapAccess.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in colmap.cc.obj
FreeImage.lib(BitmapAccess.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in colmap.cc.obj
FreeImage.lib(Plugin.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in colmap.cc.obj
FreeImage.lib(Plugin.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in colmap.cc.obj
FreeImage.lib(GetType.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in colmap.cc.obj
FreeImage.lib(GetType.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in colmap.cc.obj
FreeImage.lib(PixelAccess.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in colmap.cc.obj
FreeImage.lib(PixelAccess.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in colmap.cc.obj
...
```

It seems that in debug there is something going one with the /MD and /MT

ahojnnes commented on ports/colmap/portfile.cmake:
```diff
@@ -0,0 +1,53 @@
+    HEAD_REF dev
+)
+
+if("cuda-redist" IN_LIST FEATURES)
```
This is not correct. We still need this:

```
set(CUDA_ENABLED OFF)
vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
    cuda CUDA_ENABLED
)
```

pablospe replied:
Resolved in the last commit

pablospe committed:
Changes proposed from @ahojnne to handle the internal colmap variables `DGIT_COMMIT_ID` and `DGIT_COMMIT_DATE`

pablospe committed:
Enabling TESTS (changes sent by @ahojnnes)

JackBoosY commented on ports/colmap/portfile.cmake:
```diff
@@ -0,0 +1,85 @@
+set(CUDA_ENABLED OFF)
+set(TESTS_ENABLED OFF)
+
+if("cuda" IN_LIST FEATURES)
```
Please use `vcpkg_check_features` instead of these code, see [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_check_features.md).

ahojnnes replied:
I don't think this will work as "cuda-redist" will also have to set the same option.

pablospe replied:
Maybe there is a problem using `vcpkg_check_features()` since the features `cuda` and `cuda-redist` both enable the same flag: `CUDA_ENABLED`. And `cuda-redist` depends on `cuda`. When calling `colmap[cuda]` or `colmap[cuda-redist]` one activates the flag and the other tries to deactivate it. Any suggestion regarding this problem? Thanks!

JackBoosY replied:
@pablospe like:
```
if ("feature1" IN_LIST FEATURES AND "feature2" IN_LIST FEATURES)
        message(FATAL_ERROR "feature1 conflict with feature2, please disable one of them")
endif()
```

pablospe committed:
Updated to lastest tag to 3.6

pablospe commented:
The changes from @ras0219-msft have been incorporated, but it still says `requested changes`. How do I modify this?

ahojnnes commented on ports/colmap/vcpkg.json:
```diff
@@ -0,0 +1,49 @@
+    "gflags",
+    "glew",
+    "qt5-base",
+    "ceres",
```
This should include the "suitesparse", "lapack" features.

pablospe replied:
suiteparse is included (see [L15](https://github.com/microsoft/vcpkg/blob/eeb1df7173b9b82a05119d04ee1940d8beb2cf58/ports/colmap/vcpkg.json#L15)), and this also includes lapack (see [here](https://github.com/microsoft/vcpkg/blob/master/ports/suitesparse/CONTROL#L3))

ahojnnes commented on ports/colmap/vcpkg.json:
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "colmap",
+  "version-string": "3.6-dev.3",
```
3.6

pablospe committed:
Updating to 3.6

pablospe committed:
Ceres was missing suitesparse/lapack

JackBoosY commented:
/azp run

JackBoosY commented:
/azp run

ahojnnes commented:
@JackBoosY Seems like the CI build failures are not caused by this PR but by some unrelated Azure issues?

JackBoosY commented:
Testing...

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into colmap

JackBoosY commented on ports/colmap/portfile.cmake:
```diff
@@ -0,0 +1,85 @@
+    SHA512 9a4b4f2a49891ce8ac32ab1f2e9b859336326bada889e6de49c017a069884bb6c8ab8a2ae430d955e58fc22377c63e8fba9ce80ff959713e2878e29814d44632
+    HEAD_REF dev
+)
+
```
```suggestion

if (TRIPLET_SYSTEM_ARCH MATCHES "x86" AND ("cuda" IN_LIST FEATURES OR "cuda-redist" IN_LIST FEATURES))
    message(FATAL_ERROR "Feature cuda and cuda-redist require x64 architecture.")
endif()

```
When building colmap[cuda,cuda-redist,tests]:x86-windows
```
[63/388] cmd.exe /C "cd /D G:\12410\vcpkg\buildtrees\colmap\x86-windows-dbg\lib\PBA\CMakeFiles\pba.dir && G:\12410\vcpkg\downloads\tools\cmake-3.17.2-windows\cmake-3.17.2-win32-x86\bin\cmake.exe -E make_directory G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//. && G:\12410\vcpkg\downloads\tools\cmake-3.17.2-windows\cmake-3.17.2-win32-x86\bin\cmake.exe -D verbose:BOOL=OFF -D build_configuration:STRING=Debug -D generated_file:STRING=G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//./pba_generated_ProgramCU.cu.obj -D generated_cubin_file:STRING=G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//./pba_generated_ProgramCU.cu.obj.cubin.txt -P G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//pba_generated_ProgramCU.cu.obj.Debug.cmake"
FAILED: lib/PBA/CMakeFiles/pba.dir/pba_generated_ProgramCU.cu.obj 
cmd.exe /C "cd /D G:\12410\vcpkg\buildtrees\colmap\x86-windows-dbg\lib\PBA\CMakeFiles\pba.dir && G:\12410\vcpkg\downloads\tools\cmake-3.17.2-windows\cmake-3.17.2-win32-x86\bin\cmake.exe -E make_directory G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//. && G:\12410\vcpkg\downloads\tools\cmake-3.17.2-windows\cmake-3.17.2-win32-x86\bin\cmake.exe -D verbose:BOOL=OFF -D build_configuration:STRING=Debug -D generated_file:STRING=G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//./pba_generated_ProgramCU.cu.obj -D generated_cubin_file:STRING=G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//./pba_generated_ProgramCU.cu.obj.cubin.txt -P G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//pba_generated_ProgramCU.cu.obj.Debug.cmake"
nvcc fatal   : 32 bit compilation is only supported for Microsoft Visual Studio 2013 and earlier
CMake Error at pba_generated_ProgramCU.cu.obj.Debug.cmake:220 (message):
  Error generating
  G:/12410/vcpkg/buildtrees/colmap/x86-windows-dbg/lib/PBA/CMakeFiles/pba.dir//./pba_generated_ProgramCU.cu.obj
```

pablospe committed:
Update ports/colmap/portfile.cmake (commit suggested by JackBoosY)

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

Commit:
Merge branch 'colmap' of https://github.com/pablospe/vcpkg into colmap

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into colmap

Commit:
format manifest

strega-nil commented on ports/colmap/portfile.cmake:
```diff
@@ -0,0 +1,89 @@
+    HEAD_REF dev
+)
+
+if (TRIPLET_SYSTEM_ARCH MATCHES "x86" AND ("cuda" IN_LIST FEATURES OR "cuda-redist" IN_LIST FEATURES))
```
```suggestion
if (NOT TRIPLET_SYSTEM_ARCH STREQUAL "x64" AND ("cuda" IN_LIST FEATURES OR "cuda-redist" IN_LIST FEATURES))
```

pablospe committed:
Update ports/colmap/portfile.cmake

Co-authored-by: nicole mazzuca <mazzucan@outlook.com>

JackBoosY commented:
Waiting for merge #12766.

strega-nil commented:
/azp run
